src/Chapter3.hs

```diff
@@ -1006,6 +1161,79 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data MonsterActions = MAttack | Run
+    deriving (Show, Eq, Enum, Bounded)
+
+data Monster' = M 
+    { mHealth  :: Health
+    , mAttack  :: Attack
+    , mActions :: MonsterActions
+    } deriving Show
+
+data KnightActions 
+    = KAttack
+    | Heal
+    | SpellCast
+    deriving (Show, Eq, Enum, Bounded)
+  
+data Knight' = K
+    { kHealth  :: Health
+    , kAttack  :: Attack
+    , kDefense :: Defense
+    , kActions :: KnightActions
+    } deriving Show
+
+cycleMActions :: MonsterActions -> MonsterActions
+cycleMActions a
+    | a == maxBound   = minBound
+    | otherwise       = succ a
+
+cycleKActions :: KnightActions -> KnightActions
+cycleKActions a
+    | a == maxBound   = minBound
+    | otherwise       = succ a
```

Comment:
I see that you use such functions a lot :slightly_smiling_face: 
You can actually implement a general polymorphic function:

* http://hackage.haskell.org/package/relude-0.7.0.0/docs/Relude-Extra-Enum.html#v:next